File: src/main/java/net/runelite/cache/script/assembler/LabelVisitor.cpp
#include "LabelVisitor.h"

namespace net::runelite::cache::script::assembler
{
	using HashMap = java::util::HashMap;
	using Map = java::util::Map;
	using Logger = org::slf4j::Logger;
	using LoggerFactory = org::slf4j::LoggerFactory;
const std::shared_ptr<org::slf4j::Logger> LabelVisitor::logger = org::slf4j::LoggerFactory::getLogger(LabelVisitor::class);

	void LabelVisitor::exitInstruction(std::shared_ptr<rs2asmParser::InstructionContext> ctx)
	{
		++pos;
	}

	void LabelVisitor::enterLabel(std::shared_ptr<rs2asmParser::LabelContext> ctx)
	{
		std::wstring text = ctx->getText();
		text = text.substr(0, text.length() - 1); // remove trailing :

		logger->debug(L"Label {} is on instruction {}", text, pos);

		map.emplace(text, pos);
	}

	std::optional<int> LabelVisitor::getInstructionForLabel(const std::wstring &label)
	{
		return map[label];
	}
}

File: script/assembler/LabelVisitor.cs
﻿using System;
using System.Collections.Generic;

/*
 * Copyright (c) 2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache.script.assembler
{


	public class LabelVisitor : rs2asmBaseListener
	{
		private int pos;
		private readonly IDictionary<string, int> map = new Dictionary<string, int>();

		public override void exitInstruction(rs2asmParser.InstructionContext ctx)
		{
			++pos;
		}

		public override void enterLabel(rs2asmParser.LabelContext ctx)
		{
			string text = ctx.getText();
			text = text.Substring(0, text.Length - 1); // remove trailing :

			Console.WriteLine("Label {0} is on instruction {1}", text, pos);

			map[text] = pos;
		}

		public virtual int? getInstructionForLabel(string label)
		{
			return map[label];
		}
	}

}